### PR TITLE
Scripting Filter

### PR DESCRIPTION
**Pull request by dmnjohns**
I don't think it's perfect yet, but it's probably in a state where other people can take a look at it and provide feedback.

If we want to merge this filter, we should create a story to hold the following tasks:
- Code Review
- QE Review
- Documentation and Release Notes

### CODE REVIEW

File: repose-aggregator/artifacts/filter-bundle/src/main/application/WEB-INF/web-fragment.xml [hunk 1]
```diff
@@ -126,7 +126,7 @@
     </filter>
     <filter>
         <filter-name>valkyrie-authorization</filter-name>
-        <filter-class>org.openrepose.filters.valkyrieauthorization.ValkyrieAuthorizationFilter</filter-class>
+        <filter-class>org.openrepose.filters.scripting.ValkyrieAuthorizationFilter</filter-class>
```
Comment (Mario-Lopez) [inline repose-aggregator/artifacts/filter-bundle/src/main/application/WEB-INF/web-fragment.xml]: y, tho?

File: repose-aggregator/components/filters/scripting-filter/build.gradle [hunk 2]
```diff
@@ -0,0 +1,20 @@
+dependencies {
+    compile project(":repose-aggregator/core/core-lib")
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/build.gradle]: Why is this needed?

File: repose-aggregator/artifacts/filter-bundle/src/main/application/WEB-INF/web-fragment.xml (hunk 1 above)
Comment (dmnjohns) [inline repose-aggregator/artifacts/filter-bundle/src/main/application/WEB-INF/web-fragment.xml]: ...

I don't know what you're talking about...

File: repose-aggregator/components/filters/scripting-filter/src/main/resources/META-INF/schema/examples/scripting.cfg.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<scripting xmlns="http://docs.openrepose.org/repose/scripting/v1.0">
+    <request-script language="groovy">
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/src/main/resources/META-INF/schema/examples/scripting.cfg.xml]: Outdated example?

File: repose-aggregator/components/filters/scripting-filter/build.gradle (hunk 2 above)
Comment (dmnjohns) [inline repose-aggregator/components/filters/scripting-filter/build.gradle]: It's not. It was in the branch I started on, and I just didn't remove it.

Although, it was apparently providing jcl-over-slf4j transitively.

File: repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      scriptEngine.eval(script, bindings)
+    }
+  }
+
+  private object ScriptRunner {
+    def apply(script: String, engine: ScriptEngine): ScriptRunner = {
+      engine match {
+        case c: Compilable =>
+          logger.debug("Creating compiled script runner")
+          new CompiledScriptRunner(c.compile(script))
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala]: Just curious, was "new" necessary here?

File: repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/resources/META-INF/schema/config/bindings.xjb [hunk 5]
```diff
@@ -27,6 +27,6 @@
           schemaLocation="valkyrie-authorization.xsd">
 
     <schemaBindings>
-        <package name="org.openrepose.filters.valkyrieauthorization.config"/>
+        <package name="org.openrepose.filters.scripting.config"/>
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/resources/META-INF/schema/config/bindings.xjb]: Man, you got a few of these.

File: repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.scripting.config.ScriptingConfig
+import org.python.core.Options
+
+import scala.util.{Failure, Success, Try}
+
+@Named
+class ScriptingFilter @Inject()(configurationService: ConfigurationService)
+  extends Filter with UpdateListener[ScriptingConfig] with HttpDelegationManager with LazyLogging {
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala]: I don't think I saw the delegation manager used anywhere.

File: repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+    val request = new MockHttpServletRequest()
+    val response = new MockHttpServletResponse()
+
+    filter.doFilter(request, response, filterChain)
+
+    filterChain.getRequest shouldBe null
+    response.getStatus shouldBe HttpServletResponse.SC_INTERNAL_SERVER_ERROR
+  }
+}
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala]: I don't think I saw any tests for adding a header in the response.  Other than that, I love seeing the various languages that we know will work.

File: repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/resources/META-INF/schema/config/bindings.xjb (hunk 5 above)
Comment (dmnjohns) [inline repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/resources/META-INF/schema/config/bindings.xjb]: I must have done a replace at some point. The scripting filter started as a copy/paste of the valkyrie filter, so I must have done a replace to update all of that jank, but ended up replacing that jank where it shouldn't have been replaced.

File: repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala (hunk 4 above)
Comment (dmnjohns) [inline repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala]: It was. CompiledScriptRunner is not a case class because I didn't need to give access to the constructor parameter. Since it isn't a case class, it isn't provided an "apply" function by default. The apply function is what allows you to neglect using the "new" keyword.

I suppose I could have made it a case class, but I don't see any reason to.

Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/src/main/scala/org/openrepose/filters/scripting/ScriptingFilter.scala]: :+1:

File: repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala (hunk 7 above)
Comment (dmnjohns) [inline repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala]: Added one.

File: repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    val fakeConfigService = new FakeConfigService()
+    val filter = new ScriptingFilter(fakeConfigService)
+    val filterChain = new MockFilterChain()
+
+    val scriptingConfig = new ScriptingConfig()
+    scriptingConfig.setValue(
+      """
+        |$request.addHeader("lol", "butts")
+        |$filterChain.doFilter($request, $response)
+      """.stripMargin)
```
Comment (jennyvo) [inline repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala]: [ERROR]  found   : String
[ERROR]  required: org.openrepose.filters.scripting.config.ScriptingLanguages
[ERROR]       """.stripMargin)

Comment (jennyvo) [inline repose-aggregator/components/filters/scripting-filter/src/test/scala/org/openrepose/filters/scripting/ScriptingFilterTest.scala]: I fixed these, verify build then push

**Comment by dmnjohns** [comment]
retest this please

**Comment by dmnjohns** [comment]
retest this please

**Comment by dmnjohns** [comment]
retest this please

**Comment by dmnjohns** [comment]
It looks like Maven is struggling to find the Scripting filter itself (as in, it isn't able to load it from any of the test artifacts). I'm not overly concerned with that since Maven will go bye-bye soon enough. Also, it's not a problem I seem to have locally, so I'm not sure what's going on there.

As for the other tests, I believe they are are unrelated and expected since this is branched off of the repose-8-staging branch.

File: repose-aggregator/components/filters/scripting-filter/pom.xml [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    <properties>
+        <sonar.jacoco.itReportPath>${project.basedir}/../../target/jacoco-it.exec</sonar.jacoco.itReportPath>
+    </properties>
+
+    <dependencies>
+        <!-- Internal dependencies -->
+        <dependency>
+            <groupId>org.openrepose</groupId>
+            <artifactId>core-service-api</artifactId>
```
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/scripting-filter/pom.xml]: Should be repose-core-api.  I bet this is the reason for the Maven build failure and the extra functional test failures.

File: repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        repose = new ReposeValveLauncher(
+                reposeConfigProvider,
+                properties.getReposeJar(),
+                url,
+                properties.getConfigDirectory(),
+                reposePort
+        )
+        repose.enableDebug()
+        reposeLogSearch = new ReposeLogSearch(properties.getLogFile());
+        repose.start(killOthersBeforeStarting: false,
```
Comment (Mario-Lopez) [inline repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy]: If we're starting Repose in the test method, should we be stopping it in here, too?  Seems like we might be left with 4 instances running after this test.

Comment (dmnjohns) [inline repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy]: We could stop it in the method, but we won't be left with 4 instances running after this test in this case. The `cleanup` method will handle stopping Repose.

Side note: Deproxy should be shutdown in the `cleanupSpec` method, not the `cleanup` method.

Comment (dmnjohns) [inline repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy]: Side note: Deproxy _could_ be moved to a setupSpec and cleanupSpec block to avoid creating it for every test, but it's not a huge deal.

Comment (Mario-Lopez) [inline repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy]: Ha!  My bad.  It's always "cleanupSpec".  I didn't realize it was just "cleanup" here.

File: versions.properties [hunk 11]
```diff
@@ -100,6 +102,7 @@ org.rackspace:deproxy=0.21
 org.scala-lang:scala-library=2.11.7
 org.scala-lang:scala-xml=2.11.0-M4
 org.scalatest:scalatest_2.11=2.2.0
+org.python:jython=2.7.0
```
Comment (Mario-Lopez) [inline versions.properties]: I finally got to the bottom of the changes and thought "Yes!  I'll be able to merge this now."  Then I saw this out of order entry. :(

**Comment by Mario-Lopez** [comment]
Everything looks good.  Will merge when build passes.

**Comment by dmnjohns** [comment]
retest this please

File: repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy (hunk 10 above)
Comment (dmnjohns) [inline repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/scripting/ScriptingLanguageTest.groovy]: Yeah, we can probably change killOthersBeforeStarting to true (letting it default).

**Comment by dmnjohns** [comment]
The build passed with the expected failures yesterday! Of course, I had to rebase this morning to drop all the poms.
